### PR TITLE
fix(files_sharing): fix parsing of remote shares

### LINKED ISSUE #44318: [Bug]: Remote shares are not listed in share views
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

`ocsEntryToNode` expects that the OCS entry contains [a `mimetype` field](https://github.com/nextcloud/server/blob/7b9581129014f0c430272b8288ed9e45c0362499/apps/files_sharing/src/services/SharingService.ts#L61) and [a `file_source` field](https://github.com/nextcloud/server/blob/7b9581129014f0c430272b8288ed9e45c0362499/apps/files_sharing/src/services/SharingService.ts#L45). However, remote shares may not provide them. Due to that, the remote shares are not shown in the Share views.

Short version:
- If the remote share is pending there is no `mimetype` nor `file_source`
- If the remote share is accepted there is `mimetype` but no `file_source`; `file_source` is instead returned as `file_id`

Longer version:
Remote shares have [several base fields](https://github.com/nextcloud/server/blob/efe68d012c354598820121c0ff91dd4db937afdc/apps/files_sharing/lib/External/Manager.php#L791) which are then extended from [the information of the file based on its mountpoint](https://github.com/nextcloud/server/blob/755de1e0b57aa9c7275257142b6f852fbcd1961e/apps/files_sharing/lib/Controller/RemoteController.php#L128-L132).

However, when [pending remote shares are got they are not extended with those extra fields](https://github.com/nextcloud/server/blob/755de1e0b57aa9c7275257142b6f852fbcd1961e/apps/files_sharing/lib/Controller/RemoteController.php#L69) (as they may be still using a temporary mountpoint), so the mimetype is not available.

Once accepted a real mountpoint is set for remote shares, so the mimetype and the fileid become available. However, [the fileid is returned as `file_id`](https://github.com/nextcloud/server/blob/755de1e0b57aa9c7275257142b6f852fbcd1961e/apps/files_sharing/lib/Controller/RemoteController.php#L132) instead of `file_source`.

@skjnldsv FYI

## Steps to reproduce (scenario 1)

- Setup Nextcloud server A
- Setup Nextcloud server B
- Enable notifications app in server B
- In server A, share a file with user X of server B
- In server B, log in as user X
- Open the Files app
- Open the "Pending shares" view

### Expected result

The pending remote share is shown

###  Actual result

The pending remote share is not shown; browser console shows `Error while parsing OCS entry` with error `Missing or invalid mandatory mime`

## Steps to reproduce (scenario 2)

- Setup Nextcloud server A
- Setup Nextcloud server B
- Enable notifications app in server B
- In server A, share a file with user X of server B
- In server B, log in as user X
- Accept the remote share in the notification
- Open the Files app
- Open the "Shared with me" view

### Expected result

The accepted remote share is shown

###  Actual result

The accepted remote share is not shown; browser console shows `Trying to update/set a node without fileid`

### Expected behavior

See above

### Installation method

None

### Nextcloud Server version

master

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix https://github.com/nextcloud/server/issues/44318
- Also fix various missing data for external shares
- And also fix avatar for remote shares (+ dark avatar support)

## Pending shares
| Before | After |
|:---------:|:------:|
|||

## Files list
| Before | After |
|:---------:|:------:|
|||

### CODE REVIEW

File: core/Controller/AvatarController.php [hunk 1]
```diff
@@ -82,6 +86,9 @@
 				['Content-Type' => $avatarFile->getMimeType(), 'X-NC-IsCustomAvatar' => (int)$avatar->isCustomAvatar()]
 			);
 		} catch (\Exception $e) {
+			if ($guestFallback) {
+				return $this->guestAvatarController->getAvatarDark($userId, (string)$size);
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/Controller/AvatarController.php:89]: Why can't I ever get them right... :sweat_smile: 
@provokateurin pls help :bowing_woman:

File: apps/dav/lib/Connector/Sabre/SharesPlugin.php [hunk 2]
```diff
@@ -33,24 +34,19 @@ class SharesPlugin extends \Sabre\DAV\ServerPlugin {
...
-		IManager $shareManager
+		private Tree $tree,
+		private IUserSession $userSession,
+		private Folder $userFolder,
+		private IManager $shareManager,
 	) {
-		$this->tree = $tree;
-		$this->shareManager = $shareManager;
-		$this->userFolder = $userFolder;
 		$this->userId = $userSession->getUser()->getUID();
```
Comment (come-nc) [inline apps/dav/lib/Connector/Sabre/SharesPlugin.php:50]: I think DI can also provide userId directly. (see https://github.com/nextcloud/server/blob/7451105e3f77911525c95e88a522c7d0115d8d14/apps/user_status/lib/Controller/UserStatusController.php#L38 )

Comment (skjnldsv) [reply apps/dav/lib/Connector/Sabre/SharesPlugin.php:50]: It's legacy code, I will not touch this

**Review by artonge — APPROVED** [review]

**Review by AndyScherzinger — APPROVED** [review]

**Comment by juliushaertl** [comment]
This seems to have added running some more heavy queries twice, filed at https://github.com/nextcloud/server/issues/46546

**Pull request merged**